### ISSUE TITLE
background_downloader sometimes not returning 0 unfinished tasks, even though all tasks are complete

### ISSUE THREAD

**Issue by flying-wizzy** [781flyingdutchman/background_downloader#91]
**Describe the bug**
Hello again, while stress testing my app I found the following issue: sometimes, (1 out of 4 times maybe) when I enqueue 3 or more tasks in short succession, all tasks successfully complete, but nevertheless background_downloader still reports 1 task as unfinished.

- Disclaimer - 
Maybe this is the expected behaviour, but I would expect it to be otherwise.

**To Reproduce**
1. Enqueue several tasks right after another. 
2. Have a central register callback for status updates. 
3. Optional: Place a `await FileDownloader().allTaskIds()` in this status update callback
4. See that background_downloader sometimes doesn't log "Returning 0 unfished tasks" _after having completed all tasks_. Also, `await FileDownloader().allTaskIds()` returns a non-empty list in those cases.

**Expected behavior**
I expect all complete tasks to be removed from the list of active task ids and background_downloader to log, that there are 0 unfinished tasks left.

**Logs**
See code below to understand how logs are created. I enqueued 4 tasks in this example. All tasks do get completed. But background_downloader never returns "0 unfished tasks"
```
flutter: 113 enqueued
[Downloader] Returning 3 unfinished tasks
flutter: [113, 114, 78]
flutter: 113 running
[Downloader] Returning 3 unfinished tasks
flutter: [113, 114, 78]
[Downloader] Starting task with id 115
flutter: 115 enqueued
[Downloader] Returning 4 unfinished tasks
flutter: [113, 114, 78, 115]
flutter: 78 complete
[Downloader] Finished task with id 78
[Downloader] Returning 3 unfinished tasks
flutter: [113, 114, 115]
flutter: 114 complete
[Downloader] Finished task with id 114
[Downloader] Returning 2 unfinished tasks
flutter: [113, 115]
flutter: 115 running
[Downloader] Returning 2 unfinished tasks
flutter: [113, 115]
flutter: 115 complete
[Downloader] Returning 2 unfinished tasks
flutter: [113, 115]
[Downloader] Finished task with id 115
flutter: 113 complete
[Downloader] Returning 1 unfinished tasks
flutter: [113]
[Downloader] Finished task with id 113
```

**Code**
This is my code. I am showing the print statements so that you can follow what's producing the logs. I am basically registering callbacks for the default group only and printing the status updates.
```
void init() {
    FileDownloader().registerCallbacks(
      taskStatusCallback: downloadStatusCallback,
      taskProgressCallback: downloadProgressCallback,
    );
  }

  void downloadStatusCallback(TaskStatusUpdate taskStatusUpdate) {
    print("${taskStatusUpdate.task.taskId} ${taskStatusUpdate.status.name}");
    updateDownloadStatus(taskStatusUpdate);
    updateDefaultValueNotifiers();
  }

  void updateDefaultValueNotifiers() async {
    List<String> allTaskIds = await FileDownloader().allTaskIds();
    print(allTaskIds);
    final bool allDefaultTasksFinished = allTaskIds.isEmpty;
    updateOverallDownloadStateIfNecessary(allDefaultTasksFinished);
  }
```

I am using background_downloader 7.5.0

**Comment by 781flyingdutchman** [comment]
Thanks for stress testing!  I think the issue is in the timing, as you (understandably) check for tasks being finished when you receive the status update 'complete' - but that status update is generated by the underlying workmanager task, which is therefore still active. Per the documentation: "use `taskForId` to get the Task for the given taskId, or null if not found. Only tasks that are active (ie. not in a final state) are guaranteed to be returned, but returning a task does not guarantee that it is active". The same is true for `allTaskIds` and `isFinished` as they use they same underlying principle. The state will be *eventually* consistent, but not immediately.

A few thoughts:
1. I probably should add something to the documentation to caution using these methods when a task completes (because that is when you are most vulnerable to this happening)
2. A proper remedy is for you to keep track of the tasks to enqueue, 'tick them off' when they complete, and use that as your queue to finish the download 
3. An alternative is the use the `trackTasks` mechanism and use that to see if all tasks are finished. That should work, because the database is updated based on status updates exactly like what you do in your code
4. Introduce a small delay before checking `tasksFinished`. Tricky, because this isn't 100% fail safe (but please try and let me know if that solves it reliably)

Let me know what you think!

**Comment by flying-wizzy** [comment]
Thank you for clearing this up and for your suggestions!
What I am doing now is your proposed method 2, that seems to work just fine for now. I'll consider your 3rd proposal as well, that sounds a little easier to implement.

And as usual, thank you for your lightning fast response time 🥇

**Comment by 781flyingdutchman** [comment]
Here's what I propose to include in the README, with a code sample for method 4:
**NOTE:** Only tasks that are active (ie. not in a final state) are guaranteed to be returned or counted, but returning a task does not guarantee that it is active. This means that if you check `tasksFinished` when receiving task updates, you may get `false` even though that was the last task, as it takes a few milliseconds for the task queue to catch up. One way to remedy this is to use a `Timer`:
```agsl
// define the allTasksDoneTimer in the class containing your status checking logic
Timer? allTasksDoneTimer;

// Then, define a method to check all tasks are done, and if so, act on that.
// Call this method in your status checking logic, whenever a task completes
Future<void> checkAllTasksDone() async {
  allTasksDoneTimer?.cancel();
  final remainingTasks = await FileDownloader().allTaskIds();
  switch (remainingTasks.length) {
    case 0:
      print('all tasks are done, so do what you need to do'); 

    case 1:
      // check every seconds
      allTasksDoneTimer = Timer(const Duration(seconds: 1), checkAllTasksDone);

    default:
  }
}
```

Let me know if you think that would be helpful, or have another suggestion.

**Comment by flying-wizzy** [comment]
That looks good to me! 
What I tried - with success so far - is the following. If `remainingTasks` length is 1 and the latest `TaskStatusUpdate` is marked as `complete`, I check if that `taskId` still exists in the `remainingTasks` list. If it does, I infer, that all tasks have completed. But that solution is less generalistic compared to yours and requires `TaskStatusUpdate` information. Therefore it has to be fired from within the `downloadStatusCallback`.

A benefit could be, though, that one does not have a delay of Duration "x".


```
void downloadStatusCallback(TaskStatusUpdate taskStatusUpdate) {
    List<String> allTaskIds = await FileDownloader().allTaskIds();

    final bool lastTaskIdComplete =
        lastDefaultGroupTaskIdMarkedComplete(allTaskIds, taskStatusUpdate);

    final bool allDefaultTasksFinished =
        allTaskIds.isEmpty || lastTaskIdComplete;

    print("all tasks done, do what you need to do");
  }

bool lastDefaultGroupTaskIdMarkedComplete(
      List<String> allTaskIds, TaskStatusUpdate taskStatusUpdate) {
    if (allTaskIds.length == 1) {
      if (taskStatusUpdate.status == TaskStatus.complete &&
          allTaskIds[0] == taskStatusUpdate.task.taskId) {
        return true;
      }
    }
    return false;
  }
```

**Comment by 781flyingdutchman** [comment]
Ah, much more elegant -thanks! I am going to include something like this in the readme.

**Comment by 781flyingdutchman** [comment]
I might actually update the `tasksFinished` method to take an option `ignoreTaskId` as a parameter, and the method will return `true` id the task list is empty or contains that TaskId to ignore.  That way, you can use your initial approach (as that is by far the most intuitive).

**Comment by flying-wizzy** [comment]
Wouldn't that proposed ignore option lead to `tasksFinished` returning true, even if that task is still running? In that case I could not be certain if all tasks have actually finished or not. Or am I missing something here?

**Comment by 781flyingdutchman** [comment]
Building on your idea, you would only call it with an argument like that from within the task status update logic, and only if the task is in a 'final' state.  So it would look like this:
```
void init() {
    FileDownloader().registerCallbacks(
      taskStatusCallback: downloadStatusCallback,
      taskProgressCallback: downloadProgressCallback,
    );
  }

  void downloadStatusCallback(TaskStatusUpdate taskStatusUpdate) {
    print("${taskStatusUpdate.task.taskId} ${taskStatusUpdate.status.name}");
    updateDownloadStatus(taskStatusUpdate);
    updateDefaultValueNotifiers();
    final bool allDefaultTasksFinished = taskStatusUpdate.status.isFinalState &&  await FileDownloader().tasksFinished(ignore: taskStatusUpdate.task) ;
    updateOverallDownloadStateIfNecessary(allDefaultTasksFinished);
  }
```
You only call `tasksFinished` if the task is in a final state, and if so you exclude it from the `taskFinished` test because you know that one is already finished.  Works, right?

**Comment by flying-wizzy** [comment]
I see, that should work, yes! I didn't even know that you implemented a `isFinalState`, that's perfect 👍

**Comment by 781flyingdutchman** [comment]
Ok, published as part of 7.6.0 and 6.3.1. Note I changed the signature slightly to take the `taskId` instead of the `task` as it's easier to go from task to taskId than vice versa.

Thanks again for your help!
